zerver/tests/test_auth_backends.py

```diff
@@ -7212,5 +7216,156 @@ def test_invalid_realm_for_user_failure(self) -> None:
+    def test_ldap_group_sync(self) -> None:
+        self.init_default_ldap_database()
+
+        with self.settings(
+            LDAP_GROUP_SYNC_LIST={
+                "zulip": [
+                    "ldap_group_not_existing_in_zulip",
+                ]
+            }
+        ), self.assertLogs("zulip.ldap", "DEBUG") as ldap_log:
+            ensure_user_groups_for_ldap_sync_exist()
+            self.assertTrue(
+                UserGroup.objects.filter(name="ldap_group_not_existing_in_zulip").exists()
+            )
+
+        self.assertEqual(
+            ldap_log.output,
+            [
+                "DEBUG:zulip.ldap:creating zulip group ldap_group_not_existing_in_zulip from ldap for realm zulip"
+            ],
+        )
+
+        user_profile = self.example_user("hamlet")
+        with self.settings(LDAP_APPEND_DOMAIN="zulip.com"):
+            result = sync_user_from_ldap(user_profile, mock.Mock())
+            self.assertTrue(result)
+        self.assertTrue(user_profile.is_active)
+
+        with self.settings(
+            AUTH_LDAP_GROUP_SEARCH=LDAPSearch(
+                "ou=groups,dc=zulip,dc=com",
+                ldap.SCOPE_ONELEVEL,
+                "(objectClass=groupOfUniqueNames)",
+            ),
+            LDAP_GROUP_SYNC_LIST={
+                "zulip": [
+                    "cool_test_group",
+                ]
+            },
+            LDAP_APPEND_DOMAIN="zulip.com",
+        ), self.assertLogs("zulip.ldap", "DEBUG") as zulip_ldap_log:
+            self.assertFalse(
+                UserGroup.objects.filter(realm=get_realm("zulip"), name="cool_test_group").exists()
+            )
+
+            # Create group cool_test_group since it doesn't yet exist in Zulip
+            ensure_user_groups_for_ldap_sync_exist()
+            # Run the function a second time to test for bugs, this should noop at this point.
+            ensure_user_groups_for_ldap_sync_exist()
+
+            self.assertTrue(
+                UserGroup.objects.filter(realm=get_realm("zulip"), name="cool_test_group").exists()
+            )
+
+            user_group = UserGroup.objects.get(realm=get_realm("zulip"), name="cool_test_group")
+
+            self.assertFalse(
+                is_user_in_group(
+                    user_group,
+                    user_profile,
+                    direct_member_only=True,
+                )
+            )
+
+            sync_user_from_ldap(user_profile, mock.Mock())
+            self.assertTrue(
+                is_user_in_group(
+                    user_group,
+                    user_profile,
+                    direct_member_only=True,
+                )
+            )
+
+            # Add a user to a Zulip group that they are not member of in ldap.
+            # This implies that they should be deleted from the Zulip group
+            # upon the next sync.
+            user_profile = self.example_user("cordelia")
```

Comment:
Let's use two different variables for the two different users in the test; a common pattern is just to call the first one `hamlet` and this one, e.g., `cordelia = self.example_user("cordelia")`.

Reply:
see be9eee219929153fd20e3c1ed4d44d3e65539430